### PR TITLE
Timing issue in the TestNonblockingSubmitWithFunc

### PR DESCRIPTION
**Pull request by jdamick**
On some machines this unit would fail due to a timing issue.  Since the Invoke in the loop was using nil as the param, it would (depending on timing) lead to the workers finishing the  (w *goWorkerWithFunc) run() in the range loop over w.args as args would == nil and return. 

If an explicit time.Sleep(1 * time.Second) is added before the "	assert.EqualError(t, p.Invoke(nil), ErrPoolOverload.Error()," it is easily reproducible.

---
name: Pull request
about: Propose changes to the code
title: ''
labels: ''
assignees: ''
---

## 1. Are you opening this pull request for bug-fixs, optimizations or new feature?

bug-fix

## 2. Please describe how these code changes achieve your intention.

stated above

## 3. Please link to the relevant issues (if any).

n/a

## 4. Which documentation changes (if any) need to be made/updated because of this PR?

n/a

## 4. Checklist

- [x] I have squashed all insignificant commits.
- [x] I have commented my code for explaining package types, values, functions, and non-obvious lines.
- [x] I have written unit tests and verified that all tests passes (if needed).
- [x] I have documented feature info on the README (only when this PR is adding a new feature).
- [x] (optional) I am willing to help maintain this change if there are issues with it later.

### CODE REVIEW

**Review by panjf2000 — CHANGES_REQUESTED** [review]
Thanks for catching this issue, however, your fix doesn't resolve it completely, in fact, if you add `time.Sleep(time.Second)` after close(ch), you then get an error: `close of closed channel`, so please also fix this one.

**Comment by jdamick** [comment]
Thanks, switched to a waitgroup like the other tests to resolve that read from a closed channel.

File: ants_test.go [hunk 1]
```diff
@@ -427,24 +427,26 @@ func TestMaxBlockingSubmit(t *testing.T) {
...
 		longRunningPoolFunc(i)
-		close(ch1)
+		wg.Done()
 	}, WithNonblocking(true))
 	assert.NoError(t, err, "create TimingPool failed: %v", err)
 	defer p.Release()
+	ch := make(chan struct{})
 	for i := 0; i < poolSize-1; i++ {
-		assert.NoError(t, p.Invoke(nil), "nonblocking submit when pool is not full shouldn't return error")
+	        wg.Add(1)
```
Comment (panjf2000) [inline ants_test.go:439]: Move this line up and change it to `wg.Add(poolSize-1)`

**Review by panjf2000 — CHANGES_REQUESTED** [review]

File: ants_test.go (hunk 1 above)
Comment (jdamick) [reply ants_test.go]: done

File: ants_test.go [hunk 2]
```diff
@@ -427,24 +427,26 @@ func TestMaxBlockingSubmit(t *testing.T) {
...
+	var wg sync.WaitGroup
 	p, err := NewPoolWithFunc(poolSize, func(i interface{}) {
 		longRunningPoolFunc(i)
-		close(ch1)
+		wg.Done()
 	}, WithNonblocking(true))
 	assert.NoError(t, err, "create TimingPool failed: %v", err)
 	defer p.Release()
+	ch := make(chan struct{})
+	wg.Add(poolSize-1)
```
Comment (panjf2000) [inline ants_test.go:438]: Sorry for not giving a thoughtful comment, please just change this to `wg.Add(poolSize)` and remove the below one `wg.Add(1)`.

**Review by panjf2000 — CHANGES_REQUESTED** [review]

File: ants_test.go [hunk 3]
```diff
@@ -427,24 +427,26 @@ func TestMaxBlockingSubmit(t *testing.T) {
...
 	for i := 0; i < poolSize-1; i++ {
-		assert.NoError(t, p.Invoke(nil), "nonblocking submit when pool is not full shouldn't return error")
+		assert.NoError(t, p.Invoke(ch), "nonblocking submit when pool is not full shouldn't return error")
 	}
-	ch := make(chan struct{})
 	// p is full now.
+	wg.Add(1)
```
Comment (panjf2000) [inline ants_test.go:443]: Remove it.

File: ants_test.go (hunk 2 above)
Comment (jdamick) [reply ants_test.go]: done. either way, it's all equivalent.

**Review by panjf2000 — APPROVED** [review]

**Comment by panjf2000** [comment]
Thanks~

**Pull request merged**